### PR TITLE
[Feature/extensions] Adding ActionListener onFailure to ExtensionsOrchestrator

### PR DESCRIPTION
**Pull request by ryanbogan**
Signed-off-by: Ryan Bogan <rbogan@amazon.com>

### Description
Adding ActionListener onFailure to ExtensionsOrchestrator
 
### Issues Resolved
[61](https://github.com/opensearch-project/opensearch-sdk/issues/61)
 
### Check List
- [X ] New functionality includes testing.
  - [X ] All tests pass
- [ X] New functionality has been documented.
  - [ X] New functionality has javadoc added
- [ X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (joshpalis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:298]: Since we're adding more extension request handlers, I would suggest we refactor this method to use a switch statement rather than adding all these `else if` statements

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (joshpalis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:71]: Small change here, since all request action names from the extension to OpenSearch begin with `REQUEST_EXTENSION_` we should follow this pattern. Perhaps change to `REQUEST_EXTENSION_ACTION_LISTENER_ON_FAILURE`

**Review by joshpalis — APPROVED** [review]
LGTM! Thanks for the changes

**Review by owaiskazi19 — CHANGES_REQUESTED** [review]
We need to identify the exceptions here rather than incrementing count

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:30]: Why do we need a counter here? We are not using it anywhere

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:77]: Can we change the action to `internal:extensions.....`

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:349]: Instead of calling the listener here. We should create a separate response to handle this.

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java
Comment (owaiskazi19) [inline server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:428]: Test should be `expected, returned value`

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:30]: That's what AD does.   I'm not sure where it's used elsewhere, but I'm emulating as closely as possible to their functionality.

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:30]: Can you point any reference to the counter in AD?

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java
Comment (owaiskazi19) [reply server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:428]: I see you have marked the conversation as resolved but not push anything. Changing it back to `unresolve`.

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:30]: > We are not using it anywhere

We are using it in the getter below :-)

But seriously... I do see a counter in the MultiResponse.  Note it's using an `AtomicLong` to be thread safe, this implementation is not thread safe, and not sure it's needed in the general case.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:349]: (If it stays) More narrowly type this exception.  (If it moves, more narrowly type the new exception in the new code.)

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java
Comment (dbwiddis) [reply server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:428]: (Sometimes I click resolve when I make the changes on local before pushing... assuming that's the case here!)

File: CHANGELOG.md
Comment (dbwiddis) [reply CHANGELOG.md:15]: I'll be opening a new issue / PR later today or over the weekend to put all our feature branch changelog entries in a separate section... look for it

**Review by dbwiddis — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:30]: If I was not clear, I meant we are not using the `getter` as well anywhere.

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java
Comment (ryanbogan) [reply server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:428]: I mark it as resolved when I fix it locally

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:349]: Do you have a specific type in mind?  The actual type of the exception exists in the exception message itself since there is no way to pass over exceptions through transport service.

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:349]: > Any particular reason why? The response would just be calling the exact same line.

Because we want to follow the same code style we have for other cases as well.
The way we could have called `localNode` [here](https://github.com/opensearch-project/OpenSearch/blob/feature/extensions/server/src/main/java/org/opensearch/cluster/LocalNodeResponse.java#L29) itself but we didn't .

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:43]: Should we change the ` ExtensionRequest` here with `failureException` because `ExtensionRequest` was created to be generic for all the requests? @saratvemulapalli any thoughts here will be appreciated. Thanks

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:19]: Can we add more details for this class? Also, `ExtensionsOrchestratore` -> `ExtensionsOrchestrator`

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:43]: I'll address this in a follow-up PR

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:19]: I'll address this in a follow-up PR

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java
Comment (minalsha) [reply server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:43]: @ryanbogan do create a separate issue for tracking the follow-up PR so that we dont miss it and we can unblock this PR from merging. Thanks

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (minalsha) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:19]: @ryanbogan  do create a separate issue for tracking the follow-up PR so that we dont miss it and we can unblock this PR from merging. Thanks

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:43]: https://github.com/opensearch-project/opensearch-sdk-java/issues/135

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:19]: https://github.com/opensearch-project/opensearch-sdk-java/issues/135

**Review by owaiskazi19 — APPROVED** [review]

**Review by joshpalis — APPROVED** [review]

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:19]: (Typo) extra e at the end. Wouldn't hold up merging for this but since I've requested other changes...

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:29]: Food for thought:  in a very long running setup, this list keeps getting added to and never emptied, will eventually consume all heap space. Consider a max capacity structure, like a circular FIFO queue, LRU queue, etc.

Don't hold up merging for this but consider adding an issue to discuss.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:332]: The javadocs didn't move with the rest of the method.  As package private this may not be required per our settings but I think they bring value.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:336]: I don't see the value of temporarily using another variable here vs. the original `return new ClusterStateResponse(...)`.   Similar comment for the other cases.  But the new code is inline.

Is there a reason this is preferred?  Or did this perhaps start from a branch before the other code got trimmed down?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:351]: Please use a more narrowly typed exception here.  I'd suggest either `IllegalArgumentException` or `UnsupportedOperationException`.

Other users may want to catch and handle this and this forces them to catch all exceptions.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:357]: Can we pick a more narrow exception to send here?  Not sure which one would apply.  Do we have a custom `OpenSearchException`?

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Please use more narrow exceptions.  Other comments are suggestions you can take or leave.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:357]: We can, but the actual exception type will be in the exception message anyways.

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:19]: I fixed this in my next PR that will be up once this one merges.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:336]: I didn't change anything about this, only added a switch statement instead of if/else, since that wasn't there when I first raised the PR

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:357]: I changed it to OpenSearchException

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:29]: https://github.com/opensearch-project/opensearch-sdk-java/issues/143

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:336]: You can periodically re-sync from main.  The shorter lines and javadoc are in the diff a little lower in this file... should be a copy-paste exercise to move them.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:332]:
```suggestion
    /**
     * Handles an {@link ExtensionRequest}.
     *
     * @param extensionRequest  The request to handle, of a type defined in the {@link RequestType} enum.
     * @return  an Response matching the request.
     * @throws Exception if the request is not handled properly.
     */
    TransportResponse handleExtensionRequest(ExtensionRequest extensionRequest) throws Exception {
```

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:334]:
```suggestion
        switch (extensionRequest.getRequestType()) {
            case REQUEST_EXTENSION_CLUSTER_STATE:
                return new ClusterStateResponse(clusterService.getClusterName(), clusterService.state(), false);
            case REQUEST_EXTENSION_LOCAL_NODE:
                return new LocalNodeResponse(clusterService);
            case REQUEST_EXTENSION_CLUSTER_SETTINGS:
                return new ClusterSettingsResponse(clusterService);
```

**Review by dbwiddis — CHANGES_REQUESTED** [review]

**Comment by dbwiddis** [comment]
You need a change log entry.

Might as well fix the suggested changes and typo while you're at it.

File: server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/extensions/ExtensionActionListener.java:29]: +1

**Review by dbwiddis — APPROVED** [review]
LGTM.

**Pull request merged**
